### DISCUSSION TITLE
How can I have a after test hook( "onTestEnd" or "onTestFinish")

### DISCUSSION THREAD

**Discussion by dmanukyan** [vitest-dev/vitest#4729] [General]
I am working on developing a package to link Vitest run outcomes with TestRail. I am encountering a challenge where I cannot modify the test results following each test execution. How can I update the test results in TestRail after each test run?
If I'm correct, Vitest doesn't have a reporter hook like "onTestEnd" or "onTestFinish". Here are the hooks that I have identified.
```
interface Reporter {
    onInit?(ctx: Vitest): void;
    onPathsCollected?: (paths?: string[]) => Awaitable<void>;
    onCollected?: (files?: File[]) => Awaitable<void>;
    onFinished?: (files?: File[], errors?: unknown[]) => Awaitable<void>;
    onTaskUpdate?: (packs: TaskResultPack[]) => Awaitable<void>;
    onTestRemoved?: (trigger?: string) => Awaitable<void>;
    onWatcherStart?: (files?: File[], errors?: unknown[]) => Awaitable<void>;
    onWatcherRerun?: (files: string[], trigger?: string) => Awaitable<void>;
    onServerRestart?: (reason?: string) => Awaitable<void>;
    onUserConsoleLog?: (log: UserConsoleLog) => Awaitable<void>;
    onProcessTimeout?: () => Awaitable<void>;

}
```

**Comment by sheremet-va** [comment]
Yes, Vitest doesn't have hooks for that. All changes to test results can be found in `onTaskUpdate`. All results are updated by `state.updateTasks` method, but I don't really understand why you would need to change it post-factum.

**Reply by dmanukyan** [reply]
I require a hook to consistently update the test results in TestRail after the execution of each test case. This is necessary as the test run may be interrupted, and it's important to ensure that the most recent results are recorded in TestRail.
Can we see the test result in `onTaskUpdate` hook after **each** test execution?

**Reply by sheremet-va** [reply]
`onTaskUpdate` receives an array of "packs". The first argument is task ID. It can be a test, or it can be a suite, you need to check it yourself. The second argument is the test result. It can be undefined. The third argument is [meta object](https://vitest.dev/advanced/metadata.html#task-metadata). So in your case it would looks something like this:

```ts
function onTaskUpdate(packs: TaskResultPack[]) {
  for (const [id, result] of packs) {
    const task = this.ctx.state.idMap.get(id)
    if (task && result && task.type === 'test') {
      saveTestRailResult(task, result)
    }
  }
}
```

**Reply by dmanukyan** [reply]
Thank you for your answer. But as I understood the `onTaskUpdate` works by the number of test files and there is no way to have each test result after each test execution.

**Reply by sheremet-va** [reply]
No, it's called after each test function. There is no correlation with number of files. It's possible that there will be several results if tests were running fast

**Reply by dmanukyan** [reply]
Ok got it, thanks a lot. Yes, the reason was that the tests were running fast.
